src/render/draw_symbol.js

```diff
@@ -244,28 +258,33 @@ function drawLayerSymbols(painter, sourceCache, layer, coords, isText, translate
         const isSDF = isText || bucket.sdfIcons;
 
         const sizeData = isText ? bucket.textSizeData : bucket.iconSizeData;
+        const transformed = pitchWithMap || tr.pitch !== 0;
 
         if (!program) {
-            program = painter.useProgram(isSDF ? 'symbolSDF' : 'symbolIcon', programConfiguration);
+            program = painter.useProgram(getSymbolProgramName(isSDF, isText, bucket), programConfiguration);
             size = symbolSize.evaluateSizeForZoom(sizeData, tr.zoom);
         }
 
-        context.activeTexture.set(gl.TEXTURE0);
-
         let texSize: [number, number];
+        let texSizeIcon: [number, number] = [0, 0];
         let atlasTexture;
         let atlasInterpolation;
+        let atlasTextureIcon = null;
+        let atlasInterpolationIcon;
         if (isText) {
             atlasTexture = tile.glyphAtlasTexture;
             atlasInterpolation = gl.LINEAR;
             texSize = tile.glyphAtlasTexture.size;
-
+            if (bucket.iconsInText) {
+                texSizeIcon = tile.imageAtlasTexture.size;
+                atlasTextureIcon = tile.imageAtlasTexture;
+                const textScaled = size && size.uSize !== defaultTextSize;
+                atlasInterpolationIcon = textScaled || transformed || painter.options.rotating || painter.options.zooming ? gl.LINEAR : gl.NEAREST;
```

Comment:
@ansis @kkaefer What do you think, should I force interpolation to be gl.NEAREST? Designers would need to provide 'pixel perfect' icons for a font size being used.

Reply:
@alexshalamov we're planning to remove our use of `GL_NEAREST` in the near future anyway. See https://github.com/mapbox/mapbox-gl-js/pull/8738 for more details.